### PR TITLE
Cody: LLM-based intent detection PoC

### PR DESCRIPTION
**Pull request by vdavid**
This is something that I've wanted to do for a long time! It's a "Choose your adventure"-style, LLM-based intent detection PoC for Cody.
4-min Loom demo here: https://www.loom.com/share/1b97e6081d924033b03c99243ff3f9a8

Ways to improve this:
- Iterate on the prompt to get better choices. Btw, GPT-4 performed a lot better in my tests than our current LLM.
- Add Naman's [`getCompletionRequestMessages` logic](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/enterprise/cody/search/translateToQuery.ts?L33) to immediately get a Sourcegraph search query in case of choice "c". Then we could run that query right away, and feed _those results_ in the context for Cody. ← This would be the next most powerful add that I have in mind
- Figure out the guidelines on when to use Sourcegraph search ("c") vs. embeddings search, and add embeddings search as an option.
- I haven't actually implemented the "get all open file contents in VS Code" logic because it seemed nontrivial, especially with the limited context.

I created this as an RFC, not intending to merge this in this form.

## Test plan

This is just a PoC

## App preview:

- [Web](https://sg-web-dv-cody-llm-based-context-picking.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/cody-shared/src/intent-detector/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import vscode from 'vscode'
+import execa from 'execa'
+import path from 'path'
+
+export function getProjectName(): string | undefined {
```
Comment (vdavid) [inline client/cody-shared/src/intent-detector/helpers.ts:5]: This file only contains utility functions to get local context from VS Code

File: client/cody-shared/src/intent-detector/llm-based.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import { SourcegraphCompletionsClient } from '../sourcegraph-api/completions/client'
+
+import { IntentDetector } from '.'
+import { getCurrentBranchName, getCurrentFilePath, getOpenFilePaths, getProjectName, getRepoName } from './helpers'
+
+export class LlmBasedIntentDetector implements IntentDetector {
+    private cache = new Map<string, string>()
+
+    constructor(private completionsClient: SourcegraphCompletionsClient) {}
```
Comment (vdavid) [inline client/cody-shared/src/intent-detector/llm-based.ts:9]: This below is just a simple logic to recognize the responses "a", "b", or "c".

File: client/cody-shared/src/intent-detector/llm-based.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    public async isCodebaseContextRequired(input: string): Promise<boolean | Error> {
+        if (!this.cache.has(input)) {
+            const response = await this.determineContextNeeds(input)
+            this.cache.set(input, response)
+        }
+        return this.cache.get(input) === 'c'
+    }
+
+    private async determineContextNeeds(input: string): Promise<string> {
```
Comment (vdavid) [inline client/cody-shared/src/intent-detector/llm-based.ts:39]: This is the meat of the MVP!

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.04 kb) | 0.00% (+0.04 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/dv/cody-llm-based-context-picking-poc/compare-report.html;tab=live_object) for a full comparison between the commits 188b4fc and e20017e or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by sqs** [comment]
<kbd>*semi-automated message*</kbd> Cody's source code has moved to a new repository, https://github.com/sourcegraph/cody. If you still intend to merge this change, please reopen your PR on that new repository.

The following are some of the path mappings from the old repository (https://github.com/sourcegraph/sourcegraph) to the new repository (https://github.com/sourcegraph/cody):

- `client/cody/ -> vscode/`
- `client/cody-shared/ -> lib/shared/`
- `client/cody-ui/ -> lib/ui/`

Thank you for the PR, and sorry for the inconvenience!

**Pull request closed without merging**